TypeScript & Flux/React TodoMVC Example

bparadie opened the pull request:
Please consider adding this example to `todomvc`.

This application is a port of the [flux-todomvc](https://github.com/facebook/flux/tree/master/examples/flux-todomvc) example, which is part of Facebook's [Flux](https://github.com/facebook/flux) project.

The original example uses [JSX](http://facebook.github.io/react/docs/jsx-in-depth.html), which is a JavaScript syntax extension that looks similar to XML. 
Transforming JSX to Javascript is pretty easy as James Brantly explained in his recent React.js Conf 2015 talk [Static typing with Flow and TypeScript](http://conf.reactjs.com/schedule.html#static-typing-with-flow-and-typescript).

James Brantly's [reactconf](https://github.com/jbrantly/reactconf) uses [webpack](http://webpack.github.io/) and [ts-jsx-loader](https://github.com/jbrantly/ts-jsx-loader). This project on the other hand uses [grunt](http://gruntjs.com/) and [grunt-text-replace](https://github.com/yoniholmes/grunt-text-replace) with [react-tools](https://www.npmjs.com/package/react-tools) and [grunt-browserify](https://github.com/jmreidy/grunt-browserify) instead.

The main reason for [choosing browserify over webpack](http://blog.namangoel.com/browserify-vs-webpack-js-drama) is that browserify injects node.js polyfills like [EventEmitter](http://nodejs.org/api/events.html#events_class_events_eventemitter), which are used by the original [flux-todomvc](https://github.com/facebook/flux/tree/master/examples/flux-todomvc) example.

During the build phase all `*.tsx` and `*.ts` are first being copied to a temporary `./build` folder and then compiled to Javascript using the Typescript compiler. A second pass then intelligently replaces `React.jsx(...)` with transformed JSX code. For more details see `./grunt/replace.js`.
The build mechanics are all implemented in `./grunt/aliases.js`. The output folder is `./public`.

This project also supports minification using Google's Closure Compiler.

Here is a live demo:
http://bparadie.github.io/fluxts-todomvc

Please let me know if you have any questions, or if I need to make changes in order to comply with the rules for submitting to your repo.

arthurvr commented:
I saw you added an editorconfig and used your own code style. Please don't do so :wink:. Remove the extra editorconfig and use [our code style](https://github.com/tastejs/todomvc/blob/master/codestyle.md), please.

arthurvr commented on examples/typescript-reactflux/README.md:
```diff
@@ -0,0 +1,98 @@
+
+
+
+
```
remove all these blank lines

arthurvr commented on examples/typescript-reactflux/README.md:
```diff
@@ -0,0 +1,98 @@
+
+## Demo
+
+Please try this live demo:<br>
```
We generally don't include markup, just do an extra line or spaces.

arthurvr commented on examples/typescript-reactflux/README.md:
````diff
@@ -0,0 +1,98 @@
+```
+
+That will build `public/js/bundle.js`, which are used by `public/index.html`.
+Once you've built the project you can open `public/index.html` in your browser.
````
That workflow won't work on the live site.

arthurvr replied:
Do we really need this Grunt setup by the way?

bparadie replied:
@arthurvr The setup with `npm install` and `grunt build` is only for dev environments. What is allowed? npm scripts? I could install `grunt-cli` locally and add a post install step to `package.json`. That would leave you with one `npm install`. Would that work?

arthurvr replied:
Well, I was just seeing that many of the things we do with grunt aren't needed for us. The linting setup should be cross-app (as our code style), we don't need many steps of the build proccess, ...

arthurvr commented on examples/typescript-reactflux/.gitignore:
```diff
@@ -0,0 +1,69 @@
+src/**/*.js.map
+
+.DS_Store
+*.log
```
Nearly all these things will need to go in ones global gitignore, not here please.

bparadie replied:
@arthurvr With 'global gitignore' you mean `examples/../.gitignore', right?

All the other examples come with their own `.gitignore`:

```
./ampersand/.gitignore
./angular-dart/.gitignore
./angularjs/.gitignore
./angularjs-perf/.gitignore
./angularjs_require/.gitignore
./ariatemplates/.gitignore
./atmajs/.gitignore
./backbone/.gitignore
./backbone_marionette/.gitignore
./backbone_require/.gitignore
./batman/.gitignore
./canjs/.gitignore
./canjs_require/.gitignore
./chaplin-brunch/.gitignore
./closure/.gitignore
./derby/.gitignore
./dojo/.gitignore
./duel/.gitignore
./elm/.gitignore
./emberjs/.gitignore
./enyo_backbone/.gitignore
./exoskeleton/.gitignore
./extjs_deftjs/.gitignore
./flight/.gitignore
./jquery/.gitignore
./knockback/.gitignore
./knockoutjs/.gitignore
./knockoutjs_require/.gitignore
./meteor/.meteor/.gitignore
./plastronjs/.gitignore
./polymer/bower_components/core-selector/.gitignore
./ractive/.gitignore
./rappidjs/.gitignore
./react/.gitignore
./react-backbone/.gitignore
./serenadejs/.gitignore
./socketstream/.gitignore
./typescript-angular/.gitignore
./typescript-backbone/.gitignore
./typescript-reactflux/.gitignore
./vanilladart/.gitignore
./vanillajs/.gitignore
./vue/.gitignore
./yui/.gitignore
```
Which lines should go into the global `.gitignore`?

arthurvr replied:
Well, you should remove all junk from this gitignore file and [build up a decent one, such as the other apps have](https://github.com/tastejs/todomvc/blob/master/examples/angularjs/.gitignore).

arthurvr commented on examples/typescript-reactflux/.jshintignore:
```diff
@@ -0,0 +1 @@
+# Gruntfile.js
```
What's the deal with this file?

bparadie replied:
Used by `grunt`, needs to be in the root next to `package.json`,

arthurvr replied:
We really don't need a file with only one line which is even commented out. Please remove it. Please remove the whole app-specific linting setup as well, we have a jshint system in place which is cross-app.

arthurvr commented on examples/typescript-reactflux/externs/browserify.js:
```diff
@@ -0,0 +1,9 @@
+/**
```
with `externs` you mean it's vendor code? Should be pulled in over npm.

bparadie replied:
@arthurvr No, these extern files are for the closure compiler. Otherwise you can't compile with [ADVANCED_OPTIMIZATTION](https://developers.google.com/closure/compiler/docs/api-tutorial3). I couldn't find any extern definitions for `browserify` (happens a lot), there are no npm modules.

arthurvr commented on examples/typescript-reactflux/grunt/run.js:
```diff
@@ -0,0 +1,30 @@
+
```
strict mode would be nice

bparadie replied:
@arthurvr There is a `'use strict` in line 3.

arthurvr replied:
Whoops. You can remove this blank line, though.

arthurvr commented on examples/typescript-reactflux/grunt/run.js:
```diff
@@ -0,0 +1,30 @@
+            args: [ '<%= config.dist.server %>/app.js' ]
+        }
+    };
+};
```
missing newline

arthurvr commented on examples/typescript-reactflux/src/stores/TodoStore.ts:
```diff
@@ -0,0 +1,196 @@
+      break;
+
+    default:
+      // no op
```
can't you just leave `default:` away then?

bparadie replied:
@arthurvr Probably. But I wanted to be close to the original: https://github.com/facebook/flux/blob/master/examples/flux-todomvc/js/stores/TodoStore.js#L171

arthurvr replied:
And what reason has that even in the original example?

bparadie replied:
@arthurvr If I had to guess it's because of their `flow` tool, which might throw warnings. But I am really speculating here, because C++ compilers throw warnings if you don't implement a `default` block. Is this important? Can't we just leave that code as is? I would like to avoid changing Facebook code.

arthurvr replied:
> I would like to avoid changing Facebook code.

You will need to anyways. By the way "facebook code", is this takeover from an example they built? If so I'd love to make sure they know about this, and more importantly, that they agree on including it. Not only talking about the legal side of things but also it's just friendly to let them know and ask what they think.

Besides from that, the app is unlikely to be spec compliant so you'll need to change it anyways. 

About this line in specific, there's, really, totally no deal with including this.

arthurvr replied:
> Is this important? Can't we just leave that code as is? I would like to avoid changing Facebook code.

Also, if you're not willing to do changes to this code, you should probably close the PR. Then we're just wasting time, as quite a bunch changes will be necessary.

arthurvr commented on examples/typescript-reactflux/src/stores/TodoStore.ts:
```diff
@@ -0,0 +1,196 @@
+
+  /**
+   * Get the entire collection of TODOs.
+   * @return {object}
```
We don't do jsdoc normally.

bparadie replied:
@arthurvr The original at https://github.com/facebook/flux/blob/master/examples/flux-todomvc/js/stores/TodoStore.js#L83 uses jsdoc annotation, probably as hints for their [flow](http://flowtype.org/) tool.

arthurvr replied:
Well, generally we don't do jsdoc but if the community is strongly behind it we can consider it. WDYT @tastejs/todomvc?

arthurvr commented on examples/typescript-reactflux/src/react/ReactComponent.ts:
```diff
@@ -0,0 +1,32 @@
+}
+
+export = ReactComponent;
+
```
needless blank line.

arthurvr commented on examples/typescript-reactflux/src/dispatcher/AppDispatcher.ts:
```diff
@@ -0,0 +1,20 @@
+
+import flux = require('flux');
+var Dispatcher: flux.Dispatcher<TodoAction> = new flux.Dispatcher();
+
```
needless blank line

arthurvr commented on examples/typescript-reactflux/.editorconfig:
```diff
@@ -0,0 +1,13 @@
+insert_final_newline = true
+
+[*.md]
+trim_trailing_whitespace = false
```
please remove this file (see comment on PR)

arthurvr commented on examples/typescript-reactflux/.gitignore:
```diff
@@ -0,0 +1,69 @@
+
```
needless blank line

bparadie replied:
@arthurvr I did remove that line, but it's still showing up in the diff. Could be a `github` bug.

arthurvr commented on examples/typescript-reactflux/.jshintrc:
```diff
@@ -0,0 +1,22 @@
+  "smarttabs": true,
+  "white": true,
+  "browserify": true
+}
```
needless file

arthurvr commented on examples/typescript-reactflux/.tslintrc:
```diff
@@ -0,0 +1,53 @@
+        "check-type"
+    ]
+  }
+}
```
I would prefer not to have a typescript linting setup in this PR. We will have one at one point, but that's a separated topic as it envolves more then one app.

arthurvr commented on examples/typescript-reactflux/Gruntfile.js:
```diff
@@ -0,0 +1,54 @@
+        }
+    });
+
+    };
```
We prefer not to have app-specific build systems and rather focus on the actual framework. Is this really really needed?

arthurvr replied:
also, on this line, screwed up indentation.

bparadie replied:
@arthurvr Unfortunately I do need a framework. `react` uses `jsx` and this branch uses typescript, which means we need to x-compile from `tsx` to `jsx` to `js`. I see that other examples use npm scripts `build` and `serve`. I could add those scripts. But in the end it's really the same. Most projects need extra compilation steps. What do you suggest?

arthurvr replied:
Don't use grunt if it's not necessary. I see a good grunt system built out in here, but we don't want to clutter our examples with things that aren't totally related with the framework. You will need to remove all tasks which is useless anyways and IMHO that will make it okay to just dump some `scripts` in `package.json`.

arthurvr commented on examples/typescript-reactflux/README.md:
```diff
@@ -0,0 +1,98 @@
+
+> React is a Javascript library for building user interfaces. React uses a virtual DOM diff implementation for ultra-high performance. 
+
+> _[React - facebook.github.io/react](http://facebook.github.io/react)_
```
for the url, please do use https and include a trailing slash.

arthurvr commented on examples/typescript-reactflux/externs/react.js:
```diff
@@ -0,0 +1,18 @@
+/**
+ * @type {*}
+ */
+var MSApp;
```
Same thing. Vendor code should be pulled in over npm.

bparadie replied:
@arthurvr Same thing: those extern files are for the closure compiler. Otherwise you can't compile with ADVANCED_OPTIMIZATTION. I couldn't find any extern definitions for react (happens a lot), there are no npm modules.

arthurvr commented on examples/typescript-reactflux/grunt/aliases.js:
```diff
@@ -0,0 +1,32 @@
+
```
needless blank line

bparadie replied:
@arthurvr I did remove that line, but it's still showing up in the diff. Could be a `github` bug.

arthurvr commented on examples/typescript-reactflux/grunt/aliases.js:
```diff
@@ -0,0 +1,32 @@
+  ]);
+
+  grunt.registerTask('default', ['build']);
+};
```
do we really need this grunt file

arthurvr replied:
same thing for all others

bparadie replied:
@arthurvr Yes, this is part of the `grunt` workflow.

arthurvr commented on examples/typescript-reactflux/.editorconfig:
```diff
@@ -0,0 +1,13 @@
+# http://editorconfig.org
```
remove this comment

arthurvr commented on examples/typescript-reactflux/README.md:
```diff
@@ -0,0 +1,98 @@
+browserify injects node.js polyfills like [EventEmitter](http://nodejs.org/api/events.html#events_class_events_eventemitter),
+which are used by the original [flux-todomvc](https://github.com/facebook/flux/tree/master/examples/flux-todomvc) example.
+
+During the build phase all `*.tsx` and `*.ts` are first being copied to a temporary `./build` folder and then compiled to Javascript using the Typescript compiler.
```
> Javascript

`JavaScript`

arthurvr commented:
ping @bparadie interested in finishing things up?

bparadie commented:
@arthurvr Yes, I am. Just a bit busy right now. I am planning on going preparing another PR over this weekend. Thanks for your feedback!

bparadie commented:
@arthurvr I've made all the changes you asked me to do.
The `externs` are necessary for the closure compiler, which yields the best compression rates.
I could use a different one and then get rid of `externs`. But most devs would be interested to know how react could be optimized w/t ADVANCED_OPTIMIZATION.

The big question remains: `grunt` or not. I can could something alternative up. But something has to give. In order to generate `public/bundle.min.js` there are several steps necessary. Please advise.

Thanks again for reviewing this PR.

bparadie commented:
@arthurvr Ok, I've made a few additional changes: 
- moved `./externs` to `./src/externs`
- added `start` and `postinstall` scripts to `package.json`
- renamed `./Gruntfile` to `.grunt`
- updated `README.md`

Everything is now done via `npm install`.
Would this work?

arthurvr commented:
besides from the review comments:

* [ ] Please run our normal jshint/jscs (as in not the one you introduced) and fix all code style issues.
* [ ] We use tab indentation, this is in many places wrong right now.
* [ ] Isn't added to `learn.json` yet
* [ ] Isn't added to the homepage (`index.html`) yet.
* [ ] Doesn't run clean on the live site yet. Try pulling this branch down, opening `examples/typescript-reactflux` should directly give you the app.
* [ ] `npm install` doesn't even run clean for me.
* [ ] This needs to use npm for front-end dependencies and needs to integrate with `todomvc-app-css` and `todomvc-common` as seen with other examples.
     - Also needs a decent gitignore (all junk removed and things added similar to other examples)

arthurvr commented:
> moved ./externs to ./src/externs

Why? I think we should try to make it as clear as possible that that's third-party code.

arthurvr commented:
Another thing:

* [ ] Run all browser tests

At the moment you can't even do that because of the setup for local development being different. Another reason to not do that.

bparadie commented:
@arthurvr It looks like this it not worth my time. Thanks for the feedback. Closing.

samccone commented:
@bparadie I think we (the maintainers) are in agreement that this is a useful addition to TodoMVC and represents a common bundle of tools and languages. 

Sorry that there is quite a bit of work in adding a new project to the repo, we are actively working on making it easier for people. But we would love it if you finished this out

Thanks again for all your work so far :palm_tree:

bparadie commented:
@samccone Thanks for following up. This is a weekend project and I'll spend some time on it and submit another PR. If you guys are ok that it's going to be a bit slow I'll re-submit. Thx!

arthurvr commented:
Please re-open this PR instead of submitting a new one, but thanks!

bparadie commented:
@arthurvr Unfortunately I've already permanently deleted the repo. If it's ok, I'll open a new PR with the same title and refer to this discussion.

arthurvr commented:
Okay, sounds fine :)

bparadie commented:
Thx!

bparadie commented:
@arthurvr I have a question about `jshint`. The code is written in Typescript (see `./src/**/*.ts` files) and then cross-compiled to `bundle.js` file that I then minify to `bundle.min.js`. Minified js files never pass `jshint` and even the unminified version won't pass `jshint` because of the unminified version. I could use `tslint` (https://github.com/palantir/tslint) instead.

I looked at `typescript-angular` and that doesn't pass `jshint`:

```
js/Application.js: line 56, col 0, Identifier 'todos_1' is not in camel case.
js/Application.js: line 73, col 0, Identifier 'todos_1' is not in camel case.
js/Application.js: line 104, col 17, Expected '{' and instead saw '$location'.
js/Application.js: line 168, col 19, 'angular' is not defined.
js/Application.js: line 168, col 9, 'todomvc' is defined but never used.

js/controllers/TodoCtrl.js: line 39, col 17, Expected '{' and instead saw '$location'.

6 errors
```

The `typescript-backbone` project gives me these errors:

```
js/app.js: line 43, col 5, Missing "use strict" statement.
js/app.js: line 43, col 22, Expected '{' and instead saw 'if'.
js/app.js: line 43, col 47, Expected '{' and instead saw 'd'.
js/app.js: line 46, col 23, A constructor name should start with an uppercase letter.
js/app.js: line 52, col 5, Missing "use strict" statement.
js/app.js: line 53, col 18, 'Todo' is already defined.
js/app.js: line 84, col 5, Missing "use strict" statement.
js/app.js: line 85, col 22, 'TodoList' is already defined.
js/app.js: line 104, col 13, Expected '{' and instead saw 'return'.
js/app.js: line 119, col 5, Missing "use strict" statement.
js/app.js: line 120, col 22, 'TodoView' is already defined.
js/app.js: line 168, col 13, Expected '{' and instead saw 'this'.
js/app.js: line 191, col 5, Missing "use strict" statement.
js/app.js: line 192, col 21, 'AppView' is already defined.
js/app.js: line 277, col 5, Missing "use strict" statement.
js/app.js: line 78, col 4, 'Backbone' is not defined.
js/app.js: line 90, col 33, 'Store' is not defined.
js/app.js: line 112, col 4, 'Backbone' is not defined.
js/app.js: line 134, col 25, '_' is not defined.
js/app.js: line 134, col 36, '$' is not defined.
js/app.js: line 135, col 9, '_' is not defined.
js/app.js: line 186, col 4, 'Backbone' is not defined.
js/app.js: line 202, col 25, '$' is not defined.
js/app.js: line 206, col 9, '_' is not defined.
js/app.js: line 211, col 30, '_' is not defined.
js/app.js: line 211, col 41, '$' is not defined.
js/app.js: line 265, col 9, '_' is not defined.
js/app.js: line 273, col 4, 'Backbone' is not defined.
js/app.js: line 275, col 1, '$' is not defined.
```

In short, I have two questions:
- do cross-compiled JS files need to be jshint'ed?
- do minified JS files need to be jshint'ed?

bparadie commented:
I ran this script in examples:
```
#!/bin/bash

for f in *
do
  echo "$f `jshint $f | grep '^[1-9].* errors$'`"
done
```

Here is the output:
```
ampersand 71 errors
angular-dart 57 errors
angularjs 80 errors
angularjs-perf 52 errors
angularjs_require 76 errors
ariatemplates 105 errors
atmajs 258 errors
aurelia 308 errors
backbone 199 errors
backbone_marionette 254 errors
backbone_require 230 errors
canjs 118 errors
canjs_require 565 errors
chaplin-brunch 269 errors
closure 333 errors
componentjs 444 errors
cujo 1757 errors
derby 22 errors
dijon 118 errors
dojo 85 errors
duel 73 errors
durandal 510 errors
elm 52 errors
emberjs 226 errors
emberjs_require 
enyo_backbone 239 errors
epitome 103 errors
exoskeleton 137 errors
extjs_deftjs 218 errors
firebase-angular 103 errors
flight 307 errors
foam 52 errors
gwt 325 errors
jquery 156 errors
kendo 319 errors
knockback 453 errors
knockoutjs 103 errors
knockoutjs_require 58 errors
lavaca_require 1025 errors
meteor 121 errors
mithril 53 errors
montage 303 errors
mozart 306 errors
olives 52 errors
polymer 682 errors
puremvc 348 errors
ractive 103 errors
rappidjs 59 errors
react 154 errors
react-backbone 301 errors
reagent 52 errors
run_jshint.sh 15 errors
sammyjs 163 errors
sapui5 20 errors
serenadejs 123 errors
socketstream 9 errors
somajs 104 errors
somajs_require 110 errors
spine 224 errors
thorax 308 errors
thorax_lumbar 513 errors
troopjs_require 124 errors
typescript-angular 58 errors
typescript-backbone 228 errors
vanilladart 98 errors
vanillajs 
vue 103 errors
```

The only two examples that do *not* have any jshint errors are `emberjs_require` and `vanillajs`.
Do you guys really enforce `jshint` as a requirement for submission?

samccone commented:
@bparadie it has been an eyes only check up until this point. but as soon as https://github.com/tastejs/todomvc/commit/7e29afac9fe3d799418bd6e630bc2e83b9a6ef06 lands it will be enforced.

Obviously there are some caveats with compile to js languages but, indeed there are issues in current examples and they will be eventually all resolved.

bparadie commented:
@samccone Would you be able to provide some guidance in regards to my typescript example?
As I said the JS code is generated and not really meant to be studied. But I claim that the TS code should be clean. `jshint` does not check typescript code, but [tslint](https://github.com/palantir/grunt-tslint) does, which also provides an [npm module](https://www.npmjs.com/package/tslint).

These are the current [tslint default rules](https://github.com/palantir/grunt-tslint/blob/master/tslint.json):
```
{
  "rules": {
    "class-name": true,
    "curly": true,
    "eofline": true,
    "forin": true,
    "indent": [true, 4],
    "label-position": true,
    "label-undefined": true,
    "max-line-length": [true, 140],
    "no-arg": true,
    "no-bitwise": true,
    "no-console": [true,
        "debug",
        "info",
        "time",
        "timeEnd",
        "trace"
    ],
    "no-construct": true,
    "no-debugger": true,
    "no-duplicate-key": true,
    "no-duplicate-variable": true,
    "no-empty": true,
    "no-eval": true,
    "no-string-literal": true,
    "no-trailing-whitespace": true,
    "no-unreachable": true,
    "one-line": [true,
        "check-open-brace",
        "check-catch",
        "check-else",
        "check-whitespace"
    ],
    "quotemark": [true, "double"],
    "radix": true,
    "semicolon": true,
    "triple-equals": [true, "allow-null-check"],
    "variable-name": false,
    "whitespace": [true,
        "check-branch",
        "check-decl",
        "check-operator",
        "check-separator",
        "check-type"
    ]
  }
}
```

I think it would be fair to ask for linting of the source code, instead of linting any generated code.
In this case I should lint my TS code (as should the other two typescript examples).

Would that work for you guys?

samccone commented:
I agree linting of the source is what we want, not generated code.

As to the approach for your project, I think it would be a good idea to add a 'lint' bash file in this example folder that would run the linting over your source files. 

How does that sound?

bparadie commented:
@samccone Sounds good. I was thinking of using `gulp` or `grunt` and adding a `tslint` task instead of a lint bash file. But I can also add a bash file that runs `gulp lint`, either way.

samccone commented:
Yeah that is fine, the reason for a bash file is that some of these projects do not have JS involved in the stack at all, and being able when running the tests to just try and call `./examples/...../lint` would be ideal

i guess at that point it does not need to be bash just an executable

arthurvr commented:
I think we should hold off on bash when possible. I don't want to exclude all windows people right off. In this case, what's wrong with a package.json script?

In fact I'd like to hold off on any typescript linting till we discussed https://github.com/tastejs/todomvc/issues/347 a little more. If we add a typescript linting setup then I'd like it to be cross-project with the other typescript examples.

bparadie commented:
@arthurvr What kind of tool do you use in order to detect spaces that need to be replaced with tabs according to your guidelines?

arthurvr commented:
> What kind of tool do you use in order to detect spaces that need to be replaced with tabs according to your guidelines?

For JavaScript files, JSCS. For the rest I just use my eyes :p

bparadie commented:
@arthurvr @samccone 

I am at a point where everything looks good. But when I run the tests I am getting errors.
I tried `angularjs` and I am seeing the same errors.
What's the story with the tests?

I used `mocha allTests.js --no-timeouts --reporter spec --framework=angularjs` for this output:

```
  TodoMVC - angularjs
    When page is initially opened
      1) should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (56ms)
    New Todo
      2) should allow me to add todo items
      3) should clear text input field when an item is added
      4) should append new items to the bottom of the list
      5) should trim text input
      6) should show #main and #footer when items added
    Mark all as completed
      7) should allow me to mark all items as completed
      8) should allow me to clear the completion state of all items
      9) complete all checkbox should update state when items are completed / cleared
    Item
      10) should allow me to mark items as complete
      11) should allow me to un-mark items as complete
      12) should allow me to edit an item
      ✓ should show the remove button on hover
    Editing
      13) should hide other controls when editing
      14) should save edits on enter
      15) should save edits on blur
      16) should trim entered text
      17) should remove the item if an empty text string was entered
      18) should cancel edits on escape
    Counter
      19) should display the current number of todo items
    Clear completed button
      20) should display the correct text
      21) should remove completed items when clicked
      22) should be hidden when there are no items that are completed
    Persistence
      23) should persist its data
    Routing
      24) should allow me to display active items
      25) should respect the back button
      26) should allow me to display completed items
      27) should allow me to display all items
      28) should highlight the currently applied filter


  2 passing (44s)
  28 failing
```

bparadie commented:
@arthurvr RE: spaces to tabs...

There is `unexpand` on OSX and Linux. But it doesn't support in place replacements.
This script did the trick:

```
#!/bin/bash

# http://linux.die.net/man/1/unexpand
# find . -type file -exec ~/unexpand-in-place.sh {} \;

unexpand -a $1 > /tmp/unexpand.tmp
mv /tmp/unexpand.tmp $1
```

bparadie commented:
@arthurvr @samccone
I have created a new PR: https://github.com/tastejs/todomvc/pull/1300

bparadie commented:
@arthurvr @samccone Please disregard my questions and comments about the test framework failing.
I needed to do two things:
- start the server `cd ./tests && npm install && npm run serve` (duh!)
- add my project as an exception to `allTests.js`

Ok, I now get more reasonable results:

```
  TodoMVC - typescript-reactflux
    When page is initially opened
      ✓ should focus on the todo input field
    No Todos
      ✓ should hide #main and #footer (51ms)
    New Todo
      ✓ should allow me to add todo items (396ms)
      ✓ should clear text input field when an item is added (193ms)
      ✓ should append new items to the bottom of the list (475ms)
      ✓ should trim text input (218ms)
      ✓ should show #main and #footer when items added (211ms)
    Mark all as completed
      ✓ should allow me to mark all items as completed (476ms)
      ✓ should allow me to clear the completion state of all items (513ms)
      ✓ complete all checkbox should update state when items are completed / cleared (541ms)
    Item
      ✓ should allow me to mark items as complete (419ms)
      ✓ should allow me to un-mark items as complete (418ms)
      ✓ should allow me to edit an item (687ms)
      ✓ should show the remove button on hover
    Editing
      ✓ should hide other controls when editing (419ms)
      ✓ should save edits on enter (677ms)
      ✓ should save edits on blur (758ms)
      ✓ should trim entered text (694ms)
      ✓ should remove the item if an empty text string was entered (614ms)
      1) "after each" hook

  19 passing (37s)
  1 failing

  1) TodoMVC - typescript-reactflux "after each" hook:

      Uncaught AssertionError: 2 items expected in the todo list, 3 items observed
      + expected - actual

      -2
      +3
      
      at testOperations.js:126:12
```

arthurvr commented:
> add my project as an exception to allTests.js

What exactly needed to be added? Generally just adding the folder is enough.
